### ISSUE TITLE
Cursor isn't restoring while switching back to "edit" mode from "markdown preview"

### ISSUE THREAD

**Issue by vicrac** [Automattic/simplenote-electron#1595]
### Expected
I'd like my cursor to be restored in edit area while I switch back from "preview" to "editiing". Ideally it should be restored in last position, at worst case it should be restored anywhere (so I don't have to switch from keyboard to mouse and click to get it back, which is very inconvenient for me - I am programmer and tend to use keyboard and shortcuts only).

### Observed
While switching back from "preview" to "edit" cursor is gone and one has to restore it manually (with mouse).

### Reproduced
> 1. Go to any note.
> 2. Enable "markdown formatting", by navigating to "info" menu in the upper right corner and turning on checkbox at the bottom.
> 3. Start editing note.
> 4. Switch to "preview" mode, either by clicking "eye" icon in the upper bar or using keyboard shortcut `ctrl-shift-p`
> 5. Switch back to "edit" mode by clicking the icon again or again using the same shortcut as above.
> 6. Observe that cursor is gone (you cannot edit without resort to using mouse)

System: Linux amd64, App Version: 1.8.0 (installer Simplenote-linux-1.8.0-amd64.deb)

**Comment by SonGokussj4** [comment]
Hi. I'm waiting for this 'feature' for the whole few years I'm using SimpleNote. And it's one of the main reasons I don't Markdown Preview too often... As @vicrac said, I don't want to click back. Just `ctrl-shift-p` and the cursor should be in the same place so I can continue writing... Is this really so complicated?

**Comment by dmsnell** [comment]
We've been recently working on a refactor of some of the internals of the app which have for years made seemingly-small requests like this difficult to implement. We hope that once we finish addressing these internal issues we'll be able to move forward much more quickly than before.

That is, it can be surprisingly complicated when the wires are all criss-crossed but we're undertaking the work to clean up our data flow and make this simpler.

**Comment by SonGokussj4** [comment]
I appreciate your answer :-) Thank you. Good work, I'll be looking forward to future updates.

**Comment by vicrac** [comment]
@dmsnell I understand (although a small note on that e.g. on a issue template would be nice ;)). Do you make any assumptions on when approximately will this major refactor be done?

**Comment by dmsnell** [comment]
@vicrac I don't have any good assumptions but the work is going faster than I originally anticipated it would. it's not so much of a major refactor as it is a tedious one. once that's done, which should be soon, we'll be starting over on prioritizing the bugs/defects. this particular issue may not be the first time we attack though it might come around as a consequence of other work.

**Comment by Fraasi** [comment]
Alright, this has been bothering me too, so I thought I'd take a crack on it on a lazy monday:

You can get the current element and selection and all that with `window.getSelection().getRangeAt(0)`, but the whole editor window is built with `divs` & `spans` and you can't focus on those.
[The HTMLElement.focus() method sets focus on the specified element, if it can be focused.](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/focus),
 [more at stackoverflow on this](https://stackoverflow.com/a/1600194) 

If you put `tabindex="0"` to the element, you can get focus on it, but can't write to it.
You can focus on contenteditable element, so putting `document.querySelector('[contenteditable="true"]').focus()` on some listener where you jump back to the editor from preview, you would automatically put the cursor on the first line of the note your editing.

In short: this one is a pickle, I foresee some sweat & tears before this is solved.

The good news is, that if you use ctrl+shift+p to toggle the preview and back, you can just press `tab` to bring focus & cursor to the editors first line, no need for a mouse click.

@dmsnell Any news on how the refactoring is going?

**Comment by dmsnell** [comment]
> In short: this one is a pickle, I foresee some sweat & tears before this is solved.

It feels wonderful when someone recognizes this 😄 

The refactoring work is getting close to the finish line. In the soon-to-be-released v1.16.0 we have a number of other improvements as a result, and some bug fixes as side-effects of the updates. I'm sure there will be some regressions but we're trying to catch those.

@Fraasi I don't think we'd want to go the route of focus. We can use [scrollIntoView](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) to make sure an element is visible and we can use the [selection API](https://draftjs.org/docs/api-reference-selection-state) inside of `draft-js` in order to highlight it.

The main challenge will be tracking that caret position and updating it through note edits/changes.  As it stands `note-detail` would likely have to communicate caret position with `note-content-editor` since `note-detail` is where we flip between the rendered Markdown and the plaintext.

**Comment by Fraasi** [comment]
I haven't heard of draft.js before, but seems like it's doing what I first tried to do. You can get the caret position easily in plain js (see pic), but focusing back seem to be the pain point. Draft.js has propably solved this with some internal state or other such sorcery.

![2020-04-27_2357](https://user-images.githubusercontent.com/22402899/80421638-76c91780-88e5-11ea-9069-0a3a830c35fb.png)

Well, good luck anyway with this. Can't wait for the v1.16.0 and all the shortcut goodness 😁

**Comment by dmsnell** [comment]
Thanks! `draft-js` isn't so much the problem or the solution. The problem is that when we show a Markdown preview we render the preview HTML in a `<div>` in the place we previously showed the editor. When we swap again back to the editor it's recreating itself from scratch.

To solve this we would want to continuously communicate selection changes from the editor to the note detail. That way, when we flip back to the editor we could send it the "starting cursor" position, if you will.
